cmd/k8s-pod-headless-service-operator/run.go

```diff
@@ -0,0 +1,235 @@
+	Namespace          string `long:"namespace" env:"NAMESPACE" default:"" description:"Namespace to watch, defaults to all"`
+	Annotation         string `long:"pod-annotation" env:"POD_ANNOTATION" default:"srcd.host/create-headless-service" description:"annotation that needs to be set to 'true' for the service to be created"`
+	clientSet          *kubernetes.Clientset
+}
+
+func (r *RunCommand) ExecuteContext(ctx context.Context, args []string) error {
+	var err error
+	r.clientSet, err = r.getClientSet()
+	if err != nil {
+		return err
+	}
+	podInformer := coreinformers.NewPodInformer(r.clientSet, r.Namespace, time.Minute, cache.Indexers{})
+
+	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			err := r.setUpService(obj.(*core_v1.Pod))
+			if err != nil {
+				log.Infof("Error setting up service: %s", err)
+			}
+		},
+		UpdateFunc: func(oldObj interface{}, newObj interface{}) {
+			err := r.updateService(newObj.(*core_v1.Pod))
+			if err != nil {
+				log.Infof("Error updating service: %s", err)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			err := r.deleteService(obj.(*core_v1.Pod))
+			if err != nil {
+				log.Infof("Error deleting service: %s", err)
+			}
+		},
+	})
+
+	stop := make(chan struct{})
+	defer close(stop)
+	go podInformer.Run(stop)
+
+	log.Info("Watching pods")
+
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, syscall.SIGINT, syscall.SIGTERM)
+
+	<-sig
+	stop <- struct{}{}
+
+	return nil
+}
+
+func (r *RunCommand) getClientSet() (*kubernetes.Clientset, error) {
+	if r.clientSet != nil {
+		return r.clientSet, nil
+
+	}
+
+	var config *rest.Config
+	var err error
+	if r.KubernetesContext != "" {
+		config, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			clientcmd.NewDefaultClientConfigLoadingRules(),
+			&clientcmd.ConfigOverrides{
+				CurrentContext: r.KubernetesContext,
+			},
+		).ClientConfig()
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return kubernetes.NewForConfig(config)
+}
+
+func (r *RunCommand) hasExistingService(pod *core_v1.Pod) bool {
+	client, err := r.getClientSet()
+	if err != nil {
+		return false
+	}
+
+	_, err = client.CoreV1().Services(pod.GetNamespace()).Get(pod.GetName(), meta_v1.GetOptions{})
+	return err == nil
+}
+
+func (r *RunCommand) updateService(pod *core_v1.Pod) error {
+	log.Infof("Updating pod %s", pod.ObjectMeta.Name)
+	if pod.Annotations[r.Annotation] != "true" {
+		log.Infof("%s doesn't have annotation set, skipping", pod.ObjectMeta.Name)
+		return nil
+	}
+
+	if pod.Status.PodIP == "" {
+		log.Infof("%s doesn't have an IP yet skipping", pod.ObjectMeta.Name)
+		return nil
+	}
+
+	client, err := r.getClientSet()
+	if err != nil {
+		return err
+	}
+
+	if !r.hasExistingService(pod) {
+		log.Infof("%s has no service, creating it", pod.ObjectMeta.Name)
+		return r.setUpService(pod)
+	}
+
+	// update the pod IP
```

Comment:
Why do you assume the endpoint IP has changed in this update? Wouldn't it be better to check it before updating the whole endpoint?

Reply:
good plan